### PR TITLE
Show typing working in pinentry

### PR DESCRIPTION
**Pull request by chrisnojima**
@keybase/react-hackers

### CODE REVIEW

**Comment by chrisnojima** [comment]
a bunch of the changes are due to the protocol changing

File: go/client/ui.go [hunk 1]
```diff
@@ -717,7 +717,7 @@ func (ui SecretUI) GetPassphrase(pin keybase1.GUIEntryArg, term *keybase1.Secret
 	if term == nil {
 		term = &keybase1.SecretEntryArg{
 			Prompt:         pin.Prompt,
-			UseSecretStore: pin.Features.SecretStorage.Allow,
```
Comment (chrisnojima) [inline go/client/ui.go]: let's make the names match so we can have a more generalized algorithm for features

File: go/service/account.go [hunk 2]
```diff
@@ -45,9 +45,14 @@ func (h *AccountHandler) PassphrasePrompt(_ context.Context, sessionID int) erro
 		WindowTitle: "Passphrase needed",
 		Prompt:      "Please enter the passphrase to unlock your secret key",
 		Features: keybase1.GUIEntryFeatures{
-			SecretStorage: keybase1.SecretStorageFeature{
-				Allow: true,
-				Label: "Save in Keychain",
+			StoreSecret: keybase1.Feature{
+				Allow:   true,
+				Respond: true,
```
Comment (chrisnojima) [inline go/service/account.go]: Respond means we want the value back. We want storeSecret but don't care about showTyping

File: protocol/avdl/secret_ui.avdl [hunk 3]
```diff
@@ -37,13 +37,15 @@ protocol secretUi {
 
   string getPaperKeyPassphrase(int sessionID, string username);
 
-  record SecretStorageFeature {
+  record Feature {
```
Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: This is just a feature, not about secretStorage

File: react-native/react/native/pinentry-listeners.desktop.js [hunk 4]
```diff
@@ -63,20 +63,7 @@ export default {
...
-          result.storeSecret = false
-        }
-        /* TODO something nice like this */
-        /*
-        for (const feature in arg.features) {
-          result[feature] = arg.features[feature]
-        }
-        */
-        response.result(result)
+        response.result({passphrase: arg.passphrase, ...arg.features})
```
Comment (chrisnojima) [inline react-native/react/native/pinentry-listeners.desktop.js]: yay, generic

File: react-native/react/native/pinentry.wrapper.desktop.js [hunk 5]
```diff
@@ -17,11 +17,17 @@ class PinentryWrapper extends Component {
   }
 
   onSubmit (passphrase, features) {
-    let result = {passphrase: passphrase}
-    for (const feature in features) {
-      result[feature] = features[feature]
+    // only send the features with respond back
+    const payloadFeatures = this.state.payload && this.state.payload.features
```
Comment (chrisnojima) [inline react-native/react/native/pinentry.wrapper.desktop.js]: here we filter out just the things with respond back

File: react-native/react/native/pinentry.wrapper.desktop.js [hunk 6]
```diff
@@ -31,8 +37,8 @@ class PinentryWrapper extends Component {
   render () {
     if ('payload' in this.state) {
       return <Pinentry
-        onSubmit={this.onSubmit}
-        onCancel={this.onCancel}
+        onSubmit={(passphrase, features) => this.onSubmit(passphrase, features)}
```
Comment (chrisnojima) [inline react-native/react/native/pinentry.wrapper.desktop.js]: i think its better to just use arrows here to be explicit about the scope. before the this was some render button etc

File: react-native/react/pinentry/index.render.desktop.js [hunk 7]
```diff
@@ -9,13 +9,27 @@ export default class PinentryRender extends Component {
...
 
+  onCheck (feature, checked) {
+    this.setState({
+      features: {
+        ...this.state.features,
+        [feature]: checked
+      }
+    })
+
+    if (feature === 'showTyping') {
```
Comment (chrisnojima) [inline react-native/react/pinentry/index.render.desktop.js]: if they flip showTyping we match some internal state

**Comment by chrisnojima** [comment]
@patrickxb just to sanity check the protocol changes

File: protocol/avdl/secret_ui.avdl [hunk 8]
```diff
@@ -37,13 +37,15 @@ protocol secretUi {
...
     boolean allow;
+	boolean respond; /* if we want the value back */
```
Comment (cjb) [inline protocol/avdl/secret_ui.avdl]: Looks like indentation's bad here, a tab.

File: protocol/avdl/secret_ui.avdl [hunk 9]
```diff
@@ -37,13 +37,15 @@ protocol secretUi {
...
+  record Feature {
     boolean allow;
+    boolean respond; /* if we want the value back */
     string label;
   }
 
   record GUIEntryFeatures {
-    SecretStorageFeature secretStorage;
+    Feature storeSecret;
+    Feature showTyping;
```
Comment (patrickxb) [inline protocol/avdl/secret_ui.avdl]: When would showTyping.Allow ever equal false?  And when would the caller ever want to set showTyping.Label?

File: protocol/avdl/secret_ui.avdl (hunk 8 above)
Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: i updated the editorconfig to also match this (why i had the error)

File: protocol/avdl/secret_ui.avdl (hunk 9 above)
Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: The idea is that this control is very generic. you have a list of options you can give the window. the label is what the option is. For paperkey we might want showTyping to be false

Comment (patrickxb) [inline protocol/avdl/secret_ui.avdl]: That's different than "Allow", that's "DefaultChecked".

Comment (patrickxb) [inline protocol/avdl/secret_ui.avdl]: And I feel like GUI should be in control of labels when it can...

Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: ok i can add another flag

Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: what i'll do is if allow is there, it'll send it over the wire, with default checked as a value

Comment (patrickxb) [inline protocol/avdl/secret_ui.avdl]: The storeSecret.Allow is set by the daemon as there are specific times when we don't want to offer to store a secret.

Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: The gui is dumb in this case. If we don't like this architecture we can change it but right now this is just like an array of 'feature' which have a label and some control flags. We don't actually know (or care) what storeSecret means

Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: @patrickxb take a look. we have allow, defaultValue, readonly now

File: protocol/avdl/secret_ui.avdl [hunk 10]
```diff
@@ -37,18 +37,23 @@ protocol secretUi {
...
   }
 
   record GUIEntryArg {
     string windowTitle;
     string prompt;
+    string submitLabel;
+    string cancelLabel;
```
Comment (chrisnojima) [inline protocol/avdl/secret_ui.avdl]: per a convo w/ @MarcoPolo and @malgorithms we can have different submit buttons if we're like logging in or signing up or entering a paper key etc

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

**Comment by cjb** [comment]
:thumbsup: Looks great!

**Comment by patrickxb** [comment]
LGTM

**Pull request merged**